### DISCUSSION TITLE
Clarification related to directory entry cache

### DISCUSSION THREAD

**Discussion by syeleti-msft** [libfuse/libfuse#1058] [Q&A] [answered]
If I enable FOPEN_CACHE_DIR (struct fuse_file_info::cache_readdir) in opendir call, and read the directory, then if I open the directory for 2nd time and read it again.
Observed: I am getting the readdir call again for the 2nd opendir call.
Expected: It should be served from the directory entry cache, right?

Problem I am solving: I have a network file system and if there are multiple directory listing happening parallelly on the same directory by different processes, Can I reduce the number of readdir calls by using the cache_readdir flag?

**Comment by ashleypittman** [comment, marked as answer]
The kernel directory cache is created on `opendir`, populated with `readdir` and then published with `releasedir`, and of course invalidated on `unlink` etc.

As a result if you have two clients reading the directory contents sequentially`FOPEN_CACHE_DIR` is used correctly then the second set of `readdir` calls will come from cache.

If the second client opens the directory before the first client has closed it then kernel caching will not be used and it sounds like this is what's happening in your second case.  No you cannot enable this by setting cache_readdir flag.

**Reply by bsbernd** [reply]
@ashleypittman I might miss something, but as far as I see it, kernel enables cached readdir when previous readdir is complete

Checks in `fuse_readdir_cached` if a cached readdir should be done:
```
	if (!fi->rdc.cached) {
		/* Starting cache? Set cache mtime. */
		if (!ctx->pos && !fi->rdc.size) {
			fi->rdc.mtime = inode_get_mtime(inode);
			fi->rdc.iversion = inode_query_iversion(inode);
		}
		spin_unlock(&fi->rdc.lock);
		return UNCACHED;
	}
	/*
	 * When at the beginning of the directory (i.e. just after opendir(3) or
	 * rewinddir(3)), then need to check whether directory contents have
	 * changed, and reset the cache if so.
	 */
	if (!ctx->pos) {
		struct timespec64 mtime = inode_get_mtime(inode);

		if (inode_peek_iversion(inode) != fi->rdc.iversion ||
		    !timespec64_equal(&fi->rdc.mtime, &mtime)) {
			fuse_rdc_reset(inode);
			goto retry_locked;
		}
	}
```

Setting the cache to true is done in `fuse_readdir_cache_end`
```
static void fuse_readdir_cache_end(struct file *file, loff_t pos)
{
	struct fuse_inode *fi = get_fuse_inode(file_inode(file));
	loff_t end;

	spin_lock(&fi->rdc.lock);
	/* does cache end position match current position? */
	if (fi->rdc.pos != pos) {
		spin_unlock(&fi->rdc.lock);
		return;
	}

	fi->rdc.cached = true;
```

`fuse_readdir_cache_end` is called in `fuse_readdir_uncached`, when FUSE_READDIRPLUS / FUSE_READDIR do not return any further results - EOF.

```
	if (res >= 0) {
		if (!res) {
			struct fuse_file *ff = file->private_data;

			if (ff->open_flags & FOPEN_CACHE_DIR)
				fuse_readdir_cache_end(file, ctx->pos);
		} else if (plus) {
```

But I don't see anything related in `fuse_release` and its sub functions.

**Reply by ashleypittman** [reply]
yes, you're probably right.  The cache is made available/published when the 1st directory handle reaches the end of directory rather than when it's closed.  The point is that if other directory handles are opened and readdirs are attempted before that point there is no way today they can benefit from the kernel cache.

**Reply by bsbernd** [reply]
If there is really a use case here,  that probably could be fixed with a simple waitq. For the 2nd readdir.

**Reply by ashleypittman** [reply]
Do you mean in-kernel?  It would be a nice-to-have certainly.

From my background HPC applications will probably have one process call readdir and broadcast the results, mixed I/O workloads tend to be all over the place and will therefore populate the cache quite quickly but I've seen a few parallel applications recently where every process on a node will traverse the filesystem in parallel and that plays badly with the current directory caching semantics.  Effectively to reproduce this then launch many find processes at the same time on the same directory tree, there will be zero cache use.

**Comment by syeleti-msft** [comment]
to make the cache work, one way is to implement caching inside fs daemon, but I want to skip this if I have some alternative as this would make the logic complicated.

_If the second client opens the directory before the first client has closed it then kernel caching will not be used and it sounds like this is what's happening in your second case. No you cannot enable this by setting cache_readdir flag._

Q) In the above scenario, Can I able to access the 1st readdir directory stream/buffer from the following readdir calls from the different clients as I'm trying to reduce the number of network calls and increase the performance. any suggestions?
Reason behind asking quesiton: When the directory is having millions of files, this will make listing very slow for clients doing listing at same time as each client would make network call at the sametime.

**Comment by ashleypittman** [comment]
No, it's not possible for the 2nd client to access the cached data from the 1st client in that way.  It would be possible in theory but that would be a major re-write of the kernel fuse readdir code, it's not something you can just switch on.

For this use-case in the client I maintain I had to implement caching inside the fs daemon but as you say this is complicated.
